sql/core/src/test/scala/org/apache/spark/sql/SessionStateSuite.scala

```diff
@@ -0,0 +1,132 @@
+  }
+
+  override def afterEach(): Unit = {
+    activeSession.stop()
+  }
+
+  test("fork new session and inherit RuntimeConfig options") {
+    val key = "spark-config-clone"
+    activeSession.conf.set(key, "active")
+
+    // inheritance
+    val forkedSession = activeSession.cloneSession()
+    assert(forkedSession ne activeSession)
+    assert(forkedSession.conf ne activeSession.conf)
+    assert(forkedSession.conf.get(key) == "active")
+
+    // independence
+    forkedSession.conf.set(key, "forked")
+    assert(activeSession.conf.get(key) == "active")
+    activeSession.conf.set(key, "dontcopyme")
+    assert(forkedSession.conf.get(key) == "forked")
+  }
+
+  test("fork new session and inherit function registry and udf") {
+    activeSession.udf.register("strlenScala", (_: String).length + (_: Int))
+    val forkedSession = activeSession.cloneSession()
+
+    // inheritance
+    assert(forkedSession ne activeSession)
+    assert(forkedSession.sessionState.functionRegistry ne
+      activeSession.sessionState.functionRegistry)
+    assert(forkedSession.sessionState.functionRegistry.lookupFunction("strlenScala").nonEmpty)
+
+    // independence
+    forkedSession.sessionState.functionRegistry.dropFunction("strlenScala")
+    assert(activeSession.sessionState.functionRegistry.lookupFunction("strlenScala").nonEmpty)
+    activeSession.udf.register("addone", (_: Int) + 1)
+    assert(forkedSession.sessionState.functionRegistry.lookupFunction("addone").isEmpty)
+  }
+
+  test("fork new session and inherit experimental methods") {
+    object DummyRule1 extends Rule[LogicalPlan] {
+      def apply(p: LogicalPlan): LogicalPlan = p
+    }
+    object DummyRule2 extends Rule[LogicalPlan] {
+      def apply(p: LogicalPlan): LogicalPlan = p
+    }
+    val optimizations = List(DummyRule1, DummyRule2)
+
+    activeSession.experimental.extraOptimizations = optimizations
+
+    val forkedSession = activeSession.cloneSession()
+
+    // inheritance
+    assert(forkedSession ne activeSession)
+    assert(forkedSession.experimental ne activeSession.experimental)
+    assert(forkedSession.experimental.extraOptimizations.toSet ==
+      activeSession.experimental.extraOptimizations.toSet)
+
+    // independence
+    forkedSession.experimental.extraOptimizations = List(DummyRule2)
+    assert(activeSession.experimental.extraOptimizations == optimizations)
+    activeSession.experimental.extraOptimizations = List(DummyRule1)
+    assert(forkedSession.experimental.extraOptimizations == List(DummyRule2))
+  }
+
+  test("fork new sessions and run query on inherited table") {
+    def checkTableExists(sparkSession: SparkSession): Unit = {
+      QueryTest.checkAnswer(sparkSession.sql(
+        """
+          |SELECT x.str, COUNT(*)
+          |FROM df x JOIN df y ON x.str = y.str
+          |GROUP BY x.str
+        """.stripMargin),
+        Row("1", 1) :: Row("2", 1) :: Row("3", 1) :: Nil)
+    }
+
+    SparkSession.setActiveSession(activeSession)
```

Comment:
the setting and clearing of active session should be in a try-finally so that errors in this test (that does not clear the session) does not cascade to future tests.